flink-table/flink-sql-gateway-api/src/main/java/org/apache/flink/table/gateway/api/results/serde/LogicalTypeJsonSerializer.java

```diff
@@ -0,0 +1,210 @@
+
+        // write common fields shared by all types
+        jsonGenerator.writeStringField(FIELD_NAME_TYPE_NAME, logicalType.getTypeRoot().name());
+        jsonGenerator.writeBooleanField(FIELD_NAME_NULLABLE, logicalType.isNullable());
+        // write special fields according to type root
+        switch (logicalType.getTypeRoot()) {
+            case BOOLEAN:
+            case TINYINT:
+            case SMALLINT:
+            case INTEGER:
+            case BIGINT:
+            case FLOAT:
+            case DOUBLE:
+            case DATE:
+            case NULL:
+                break;
+            case CHAR:
+                jsonGenerator.writeNumberField(
+                        FIELD_NAME_LENGTH, ((CharType) logicalType).getLength());
+                break;
+            case VARCHAR:
+                jsonGenerator.writeNumberField(
+                        FIELD_NAME_LENGTH, ((VarCharType) logicalType).getLength());
+                break;
+            case BINARY:
+                jsonGenerator.writeNumberField(
+                        FIELD_NAME_LENGTH, ((BinaryType) logicalType).getLength());
+                break;
+            case VARBINARY:
+                jsonGenerator.writeNumberField(
+                        FIELD_NAME_LENGTH, ((VarBinaryType) logicalType).getLength());
+                break;
+            case DECIMAL:
+                jsonGenerator.writeNumberField(
+                        FIELD_NAME_PRECISION, ((DecimalType) logicalType).getPrecision());
+                jsonGenerator.writeNumberField(
+                        FIELD_NAME_SCALE, ((DecimalType) logicalType).getScale());
+                break;
+            case TIMESTAMP_WITHOUT_TIME_ZONE:
+                jsonGenerator.writeNumberField(
+                        FIELD_NAME_PRECISION, ((TimestampType) logicalType).getPrecision());
+                break;
+            case TIMESTAMP_WITH_LOCAL_TIME_ZONE:
+                jsonGenerator.writeNumberField(
+                        FIELD_NAME_PRECISION,
+                        ((LocalZonedTimestampType) logicalType).getPrecision());
+                break;
+            case MAP:
+                serializeMap((MapType) logicalType, jsonGenerator);
+                break;
+            case ARRAY:
+                serializeCollection(((ArrayType) logicalType).getElementType(), jsonGenerator);
+                break;
+            case MULTISET:
+                serializeCollection(((MultisetType) logicalType).getElementType(), jsonGenerator);
+                break;
+            case ROW:
+                serializeRow((RowType) logicalType, jsonGenerator);
+                break;
+            case RAW:
+                if (logicalType instanceof RawType) {
+                    serializeRaw((RawType<?>) logicalType, jsonGenerator);
+                    break;
+                }
+                // fall through
+            default:
+                throw new ValidationException(
```

Comment:
I think it's better to use `UnsupportedException`.

Reply:
Changed.